File: 11.动态规划/17.一和零/474.一和零.py
#
# @lc app=leetcode.cn id=474 lang=python3
#
# [474] 一和零
#
from mytools import *
# @lc code=start
class Solution:
    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        dp = [[[0 for _ in range(n + 1)] for _ in range(m + 1)] for _ in range(len(strs) + 1)]
        for i in range(1, len(strs) + 1):
            zero, one = strs[i - 1].count('0'), strs[i - 1].count('1')
            for j in range(m + 1):
                for k in range(n + 1):
                    dp[i][j][k] = dp[i - 1][j][k]
                    if zero <= j and one <= k:
                        dp[i][j][k] = max(dp[i][j][k], dp[i - 1][j - zero][k - one] + 1)
        return dp[len(strs)][m][n]
# @lc code=end
class Solution:
    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        dp = [[0] * (n + 1) for _ in range(m + 1)]
        for s in strs:
            zero = s.count('0')
            one = s.count('1')
            for j in range(m, zero - 1, -1):
                for k in range(n, one - 1, - 1):
                    dp[j][k] = max(dp[j][k], dp[j - zero][k - one] + 1)
        return dp[m][n]

File: 11.动态规划/17.一和零/474.一和零.cpp
/*
 * @lc app=leetcode.cn id=474 lang=cpp
 *
 * [474] 一和零
 */

// @lc code=start
#pragma once
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
#pragma GCC optimize("Ofast,no-stack-protector,unroll-loops")
#define ALL(v) v.begin(), v.end()
#define For(i, _) for (int i = 0, i##end = _; i < i##end; ++i)
#define FOR(i, _, __) for (int i = _, i##end = __; i < i##end; ++i)
#define Rep(i, _) for (int i = (_); i >= 0; --i)
#define REP(i, __, _) for (int i = (__), i##end = _; i >= i##end; --i)
typedef long long ll;
typedef unsigned long long ull;
#define V vector
#define pb push_back
#define pf push_front
#define qb pop_back
#define qf pop_front
#define eb emplace_back
typedef pair<int, int> pii;
typedef pair<ll, int> pli;
#define fi first
#define se second
const int dir[4][2] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
const int inf = 0x3f3f3f3f;
const ll infl = 0x3f3f3f3f3f3f3f3fll;
const int mod = 1e9 + 7;
template <class T>
inline bool ckmin(T &x, const T &y) { return x > y ? (x = y, true) : false; }
template <class T>
inline bool ckmax(T &x, const T &y) { return x < y ? (x = y, true) : false; }
int init = []()
{ ios::sync_with_stdio(false); cin.tie(nullptr); return 0; }();

class Solution {
public:
    int findMaxForm(vector<string>& strs, int m, int n) {
        V<V<int>> dp(m + 1, V<int>(n + 1, 0));
        auto get = [&] (string s) -> pii
        {
            int zero = 0, one = 0;
            for (auto& ch : s)
            {
                if (ch == '0') zero ++;
                else one ++;
            }
            return {zero, one};
        };
        for (auto& s : strs)
        {
            auto [zero, one] = get(s);
            REP(i, m, zero) REP(j, n, one)
            {
                dp[i][j] = max(dp[i][j], dp[i - zero][j - one] + 1);
            }
        }
        return dp[m][n];
    }
};
// @lc code=end

